homeassistant/components/plaato/__init__.py

```diff
@@ -59,32 +75,116 @@
     extra=vol.ALLOW_EXTRA,
 )
 
+SCAN_INTERVAL = timedelta(minutes=10)
 
-async def async_setup(hass, hass_config):
+
+async def async_setup(hass: HomeAssistant, config: dict):
     """Set up the Plaato component."""
     return True
 
 
-async def async_setup_entry(hass, entry):
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Configure based on config entry."""
     if DOMAIN not in hass.data:
         hass.data[DOMAIN] = {}
 
-    webhook_id = entry.data[CONF_WEBHOOK_ID]
-    hass.components.webhook.async_register(DOMAIN, "Plaato", webhook_id, handle_webhook)
+    use_webhook = entry.data.get(CONF_USE_WEBHOOK, False)
+
+    if use_webhook:
+        setup_webhook(hass, entry)
+    else:
+        await async_setup_coordinator(hass, entry)
 
-    hass.async_create_task(hass.config_entries.async_forward_entry_setup(entry, SENSOR))
+    for platform in PLATFORMS:
+        if entry.options.get(platform, True):
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(entry, platform)
+            )
 
+    entry.add_update_listener(async_reload_entry)
     return True
 
 
-async def async_unload_entry(hass, entry):
+def setup_webhook(hass: HomeAssistant, entry: ConfigEntry):
+    """Init webhook based on config entry."""
+    if entry.data[CONF_WEBHOOK_ID] is not None:
+        webhook_id = entry.data[CONF_WEBHOOK_ID]
+        device_name = entry.data[CONF_DEVICE_NAME]
+        hass.components.webhook.async_register(
+            DOMAIN, f"{DOMAIN}.{device_name}", webhook_id, handle_webhook
+        )
+    else:
+        raise InvalidStateError
+
+
+async def async_setup_coordinator(hass: HomeAssistant, entry: ConfigEntry):
+    """Init auth token based on config entry."""
+    auth_token = entry.data.get(CONF_TOKEN)
+    device_type = entry.data.get(CONF_DEVICE_TYPE)
+
+    coordinator = PlaatoCoordinator(hass, auth_token, device_type)
+    await coordinator.async_refresh()
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```

Comment:
It looks like there are different object types being stored in the save `dict`.  Would you please give them their own keys
`hass.data[DOMAIN][entry.entry_id]` -> `hass.data[DOMAIN][CONFIG][entry.entry_id]`
`hass.data[DOMAIN][device_id]` ->  `hass.data[DOMAIN][DEVICES][entry.entry_id]`

Reply:
@bdraco Is it possible to get hold of the ConfigEntry in the method that handles the webhook? `handle_webhook`
If not do you have any suggestion on how to refactor to be able to do you suggested change?

Reply:
and what would `device` be in you suggestion?